File: SampleAppObjC/SampleAppObjC/ContainerViewController.m
//
//  ContainerViewController.m
//  WhiteLabelDemoApp
//
//  Created by ani on 7/11/16.
//  Copyright © 2016 LaunchKey. All rights reserved.
//

#import "ContainerViewController.h"

@interface ContainerViewController ()

@end

@implementation ContainerViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    //Navigation Bar Buttons
    UIBarButtonItem *rightItemRefresh = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"NavRefresh"] style:UIBarButtonItemStyleDone target:self action:@selector(btnRefreshPressedNav:)];
    [rightItemRefresh setAccessibilityIdentifier:@"checkForRequests_refresh"];
    [self.navigationItem setRightBarButtonItems:[NSArray arrayWithObjects:rightItemRefresh, nil]];

    self.navigationItem.title =  @"Auth Request View";
    
    dispatch_async(dispatch_get_main_queue(), ^{
        
        [[AuthRequestManager sharedManager] checkForPendingAuthRequest:self.navigationController withCompletion:^(NSString *message, NSError *error)
         {
             if(error != nil)
             {
                 NSLog(@"Error: %@", error);
                 if(error.code == 401)
                     [self deviceUnlinked];
                 
                 if([error.localizedFailureReason isEqualToString:@"Authorization already has response!"])
                 {
                     UIAlertView *alert = [[UIAlertView alloc] initWithTitle:[NSString stringWithFormat:@"Authorization already has a response"]
                                                                     message:nil
                                                                    delegate:self
                                                           cancelButtonTitle:@"OK"
                                                           otherButtonTitles:nil];
                     
                     [alert show];
                 }
             }
             else
                 NSLog(@"Message: %@", message);
         }];
    });
}

-(void)viewWillAppear:(BOOL)animated
{
    [[NSNotificationCenter defaultCenter] removeObserver:self name:requestApproved object:nil];
    [[NSNotificationCenter defaultCenter] removeObserver:self name:requestDenied object:nil];
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(requestApproved) name:requestApproved object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(requestDenied) name:requestDenied object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(deviceUnlinked) name:deviceUnlinked object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(requestReceived) name:requestReceived object:nil];
}

-(void)viewWillDisappear:(BOOL)animated
{
    [[NSNotificationCenter defaultCenter] removeObserver:self name:deviceUnlinked object:nil];
    [[NSNotificationCenter defaultCenter] removeObserver:self name:requestReceived object:nil];
}

-(void)requestApproved
{
    // This will be called when an auth request has been approved... Add any custom UI here
}

-(void)requestDenied
{
    // This will be called when an auth request has been denied... Add any custom UI here
}

-(void)deviceUnlinked
{
    dispatch_async(dispatch_get_main_queue(), ^{
        [self.navigationController popViewControllerAnimated:NO];
    });
}

-(void)requestReceived
{
    dispatch_async(dispatch_get_main_queue(), ^{
        
        [[AuthRequestManager sharedManager] checkForPendingAuthRequest:self.navigationController withCompletion:^(NSString *message, NSError *error)
         {
             if(error != nil)
             {
                 NSLog(@"Error: %@", error);
                 
                 if([error.localizedFailureReason isEqualToString:@"Authorization already has response!"])
                 {
                     UIAlertView *alert = [[UIAlertView alloc] initWithTitle:[NSString stringWithFormat:@"Authorization already has a response"]
                                                                     message:nil
                                                                    delegate:self
                                                           cancelButtonTitle:@"OK"
                                                           otherButtonTitles:nil];
                     
                     [alert show];
                 }
             }
             else
                 NSLog(@"Message: %@", message);
         }];
    });
}

#pragma mark - Menu Methods
-(void)btnRefreshPressedNav:(id)sender
{
    dispatch_async(dispatch_get_main_queue(), ^{
        
        [[AuthRequestManager sharedManager] checkForPendingAuthRequest:self.navigationController withCompletion:^(NSString *message, NSError *error)
         {
             if(error != nil)
             {
                 NSLog(@"Error: %@", error);
                 
                 if([error.localizedFailureReason isEqualToString:@"Authorization already has response!"])
                 {
                     UIAlertView *alert = [[UIAlertView alloc] initWithTitle:[NSString stringWithFormat:@"Authorization already has a response"]
                                                                     message:nil
                                                                    delegate:self
                                                           cancelButtonTitle:@"OK"
                                                           otherButtonTitles:nil];
                     
                     [alert show];
                 }
             }
             else
                 NSLog(@"Message: %@", message);
         }];
    });
}

@end

File: SampleAppSwift/SampleAppSwift/ContainerViewController.swift
//
//  ContainerViewController.swift
//  WhiteLabelDemoAppSwift
//
//  Created by ani on 7/22/16.
//  Copyright © 2016 LaunchKey. All rights reserved.
//

import Foundation

class ContainerViewController: UIViewController
{
    override func viewDidLoad() {
        
        super.viewDidLoad()
        
        self.title = "Auth Request View"
        
        //Navigation Bar Buttons
        self.navigationItem.leftBarButtonItem = UIBarButtonItem(image: UIImage(named: "NavBack"), style: UIBarButtonItem.Style.plain, target: self, action: #selector(ContainerViewController.back))
        self.navigationItem.leftBarButtonItem?.tintColor = UIColor.white
        let rightBarItem = UIBarButtonItem(image: UIImage(named: "NavRefresh"), style: UIBarButtonItem.Style.plain, target: self, action: #selector(ContainerViewController.refresh))
        rightBarItem.accessibilityIdentifier = "checkForRequests_refresh"
        self.navigationItem.rightBarButtonItem = rightBarItem
        self.navigationItem.rightBarButtonItem?.tintColor = UIColor.white
        
        AuthRequestManager.shared().check(forPendingAuthRequest: self.navigationController, withCompletion: { (message,error) in
            if let requestError = error
            {
                print("Auth-request error: \(requestError)")
                if let reason = (requestError as NSError).localizedFailureReason, reason == "Authorization already has response!"
                {
                    let alert = UIAlertView()
                    alert.title = "Authorization already has a response"
                    alert.addButton(withTitle: "OK")
                    alert.show()
                }
            }
            else if let requestMessage = message
            {
                print("Auth-request message: \(requestMessage)")
            }
        })
        
        NotificationCenter.default.addObserver(self, selector:#selector(requestIsApproved), name: NSNotification.Name(rawValue: requestApproved), object: nil)
        NotificationCenter.default.addObserver(self, selector:#selector(requestIsDenied), name: NSNotification.Name(rawValue: requestDenied), object: nil)
        NotificationCenter.default.addObserver(self, selector:#selector(requestIsReceived), name: NSNotification.Name(rawValue: requestReceived), object: nil)

    }
    
    override func didReceiveMemoryWarning()
    {
        super.didReceiveMemoryWarning()
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        
        let noficationName = Notification.Name(requestReceived)
        NotificationCenter.default.removeObserver(self, name: noficationName, object: nil)
    }
    
    @objc func requestIsApproved()
    {
        // This will be called when an auth request has been approved... Add any custom UI here
    }
    
    @objc func requestIsDenied()
    {
        // This will be called when an auth request has been denied... Add any custom UI here
    }
    
    @objc func requestIsReceived()
    {
        // This will be called when the device has received a pending Auth Request
        
        AuthRequestManager.shared().check(forPendingAuthRequest: self.navigationController, withCompletion: { (message,error) in
            if let requestError = error
            {
                print("Auth-request error: \(requestError)")
                if let reason = (requestError as NSError).localizedFailureReason, reason == "Authorization already has response!"
                {
                    let alert = UIAlertView()
                    alert.title = "Authorization already has a response"
                    alert.addButton(withTitle: "OK")
                    alert.show()
                }
            }
            else if let requestMessage = message
            {
                print("Auth-request message: \(requestMessage)")
            }
        })
    }
    
    @IBAction func back()
    {
        if let navController = self.navigationController
        {
            navController.popViewController(animated: true)
        }
    }
    
    @IBAction func refresh()
    {
        AuthRequestManager.shared().check(forPendingAuthRequest: self.navigationController, withCompletion: { (message,error) in
            if let requestError = error
            {
                print("Auth-request error: \(requestError)")
                if let reason = (requestError as NSError).localizedFailureReason, reason == "Authorization already has response!"
                {
                    let alert = UIAlertView()
                    alert.title = "Authorization already has a response"
                    alert.addButton(withTitle: "OK")
                    alert.show()
                }
            }
            else if let requestMessage = message
            {
                print("Auth-request message: \(requestMessage)")
            }
        })
    }
}
